### PR TITLE
Support for unquotings in Types and more

### PR DESCRIPTION
**Pull request by Dveim**

### CODE REVIEW

File: scalameta/src/main/scala/scala/meta/Trees.scala [hunk 1]
```diff
@@ -170,7 +175,7 @@ package scala.meta.internal.ast {
     @ast class While(expr: Term, body: Term) extends Term
     @ast class Do(body: Term, expr: Term) extends Term
     @ast class For(enums: Seq[Enumerator] @nonEmpty, body: Term) extends Term with Scope {
-      require(enums.head.isInstanceOf[Enumerator.Generator])
+      require(enums.head.isInstanceOf[Enumerator.Generator] || enums.head.isInstanceOf[Enumerator.Generator.Quasi])
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: Do we really need this? Isn't `Enumerator.Generator.Quasi` an `Enumerator.Generator`?

File: scalameta/src/main/scala/scala/meta/Trees.scala [hunk 2]
```diff
@@ -195,7 +200,7 @@ package scala.meta.internal.ast {
       // require(keywords.contains(value) ==> isBackquoted)
     }
     @ast class Select(qual: Term.Ref, name: Type.Name) extends Type.Ref with Pat.Type.Ref {
-      require(qual.isPath || qual.isInstanceOf[Term.Super])
+      require(qual.isPath || qual.isInstanceOf[Term.Super] || qual.isInstanceOf[Term.Quasi])
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: Why do we allow `qual` to be `Term.Quasi`? Why not make it `become` a `Term.Ref.Quasi`?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 3]
```diff
@@ -562,16 +563,22 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
+          implicitly[AstMetadata[T]].quasi(0, unquote.tree)
+        } else ahead {
+          implicitly[AstMetadata[T]].quasi(0, unquote.tree)
+        }
       case _ =>
         unreachable(debug(token))
     }
   }
 
+  def unquote[T <: Tree : AstMetadata]: T = unquote[T](advance = true) // to write `unquote[T]` without round braces
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Round braces are called parentheses :)

File: scalameta/src/main/scala/scala/meta/Trees.scala (hunk 1 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: It is, removed.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 4]
```diff
@@ -660,9 +669,12 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
     }
   }
 
-  def makeTuplePatParens(bodyf: => List[Pat.Arg]): Pat = autoPos {
-    val body = inParens(if (token.is[`)`]) Nil else bodyf.map(_.require[Pat]))
-    makeTuple[Pat](body, () => Lit.Unit(), Pat.Tuple(_))
+  def makeTuplePatParens(): Pat = {
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why does this function no longer take an argument?

Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Also, is it going to work for `p"()"`?

Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: I moved it inside `makeTuplePatParens`, since there was only one explicit usage.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 5]
```diff
@@ -1961,7 +2005,7 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
       case _: Xml.Start =>
         xmlPat()
       case _: `(` =>
-        makeTuplePatParens(noSeq.patterns())
+        makeTuplePatParens()
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: This change is, I guess, the reason for the change in the signature of `makeTuplePatParens`. So I'm just trying to understand the motivation. The thing is that in the current shape `makeTuplePatParens` is different from other make-tuple functions, so it'd be nice to have a good reason for such a discrepancy.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 6]
```diff
@@ -969,7 +998,9 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
       if (advance) next()
       res
     case token: Unquote =>
-      unquote[T]
+      unquote[T](advance)
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Isn't this equivalent to `if (advance) unquote[T] else ahead(unquote[T])`? If yes, then why have a special version of `unquote` and `ellipsis`, if we can localize this special case to a single function?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 7]
```diff
@@ -1682,10 +1708,23 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
...
-    val cases = caseSeparated { caseClause() }
+    val cases = new ListBuffer[Case]
+    while (token.is[CaseIntro] || token.is[Ellipsis]) {
+      if (token.is[Ellipsis]) {
+        cases += ellipsis(1, unquote[Case], accept[`case`])
+        if (token.is[StatSep]) next()
+      } else if (token.is[`case`] && ahead(token.is[Unquote])) {
+        next()
+        cases += unquote[Case]
+        if (token.is[StatSep]) next()
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why is this `StatSep`-based skipping necessary here?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 8]
```diff
@@ -1682,10 +1708,23 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
    *  }}}
    */
   def caseClauses(): List[Case] = {
-    val cases = caseSeparated { caseClause() }
+    val cases = new ListBuffer[Case]
+    while (token.is[CaseIntro] || token.is[Ellipsis]) {
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Where did this `CaseIntro` come from?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 6 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: About method `name`: I thought that would be strange, if `advance` didn't have influence on both cases, `Unquote` and `Ellipsis`.
And about method `unquote`: since I have had to add that `advance: Boolean = true`, compiler did not allow simple `unquote[T]` elsewhere, only with parentheses. So I added special version of `unquote`, to preserve old syntax.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 9]
```diff
@@ -1661,19 +1692,14 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
 
   def caseClause(): Case = atPos(in.prevTokenPos, auto) {
     require(token.isNot[`case`] && debug(token))
-    token match {
-      case token: Unquote if token.prev.is[Ellipsis] =>
-        unquote[Case]
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Would it be possible to retain this pattern match clause in `caseClause` (more precisely, changing it to `case token: Unquote if token.prev.is[Case]`) and then migrating `caseClauses` back to `caseSeparated`?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 7 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: To allow `;` after cases. 
`..case $as; ..case $bs`

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 8 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: It is used to handle situations like
`val q"$expr match { case bar => baz; ..case $casez; case q => w}" = q"foo match { case bar => baz; case _ => foo ; case q => w }"`

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala [hunk 10]
```diff
@@ -2911,7 +2966,7 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
    */
   def blockStatSeq(): List[Stat] = {
     val stats = new ListBuffer[Stat]
-    while (!token.is[StatSeqEnd] && !token.is[CaseDefEnd]) {
+    while (!token.is[StatSeqEnd] && !token.is[CaseDefEnd] && !ahead(token.is[`case`])) {
```
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: So `{ case class ... }` isn't going to work?

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala [hunk 11]
```diff
@@ -297,9 +297,11 @@ class ErrorSuite extends FunSuite {
       val tpes = List(q"T")
       q"expr: ..$tpes"
     """) === """
-      |<macro>:5: identifier expected but ellipsis found
+      |<macro>:5: type mismatch when unquoting;
+      | found   : List[scala.meta.Term.Name]
+      | required: scala.collection.immutable.Seq[scala.meta.Type]
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: Hmm. Why did this error message change?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 10 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Sorry, cannot understand this example. Why it should?

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala (hunk 11 above)
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: Also, what's it mean really? What if the user passed a sequence of types there? Is it going to compiler?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 12]
```diff
@@ -192,9 +183,19 @@ class QuasiquoteSuite extends FunSuite {
     assert(tpe.show[Syntax] === "Double")
   }
 
-//  test("q\"$expr: ..@$expr\"") {
-//    val q"$a: ..@$b" = q"foo: @bar @baz"
-//  }
+  test("1 q\"$expr: ..@annots\"") {
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Shouldn't there be a dollar in the name of the test?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 13]
```diff
@@ -192,9 +183,19 @@ class QuasiquoteSuite extends FunSuite {
...
+    val q"$exprr: @q ..@$annotz @$ar" = q"foo: @q @w @e @r"
+    assert(exprr.show[Syntax] === "foo")
+    assert(annotz.toString === "List(@w, @e)")
+    assert(annotz(0).show[Syntax] === "@w")
+    assert(annotz(1).show[Syntax] === "@e")
+    assert(ar.show[Syntax] === "@r")
+  }
+
+  test("2 q\"$expr: ..@annots\"") {
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: And here as well.

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 14]
```diff
@@ -220,20 +235,46 @@ class QuasiquoteSuite extends FunSuite {
...
-//  }
+  test("3 q\"$expr match { ..case $cases }\"") {
+    val q"$expr match { ..case $casez }" = q"foo match { case bar => baz; case _ => foo }"
+    assert(expr.show[Syntax] === "foo")
+    assert(casez.toString === "List(case bar => baz, case _ => foo)")
+    assert(casez(0).show[Syntax] === "case bar => baz")
+    assert(casez(1).show[Syntax] === "case _ => foo")
+  }
+
+  // todo change to expropt (and test it) after issue #199 resolved
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Could you change the comment to say `TODO: ...` to be in line with other todos in the repo?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 9 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: But `case x => y; case q => w` is not case-separated. There is one extra `case`.

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 15]
```diff
@@ -276,16 +313,65 @@ class QuasiquoteSuite extends FunSuite {
...
+    val b = enumerator"b <- bs"
+    val ab = List(a,b)
+    assert(q"for (..$ab) foo".show[Syntax] === "for (a <- as; b <- bs) foo")
+  }
 
-//  test("q\"for (..$enumerators) yield $expr\"") {
-//    val q"for (..$enumerators) yield $expr" =  q"for (x <- xs; y <- ys) yield foo(x, y)" // fixme test is broken, so even does not compile
+//  test("3 q\"for (..$enumerators) $expr\"") {
+//    val q"for (a <- as; if $cond; ..$enums) bar" = q"for (a <- as; if foo; b <- bs) bar" // fixme does not compile
 //  }
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Could you link https://github.com/scalameta/scalameta/issues/203 in a comment to this test?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 16]
```diff
@@ -276,16 +313,65 @@ class QuasiquoteSuite extends FunSuite {
...
+  }
+
+  test("3 q\"new { ..$stat } with ..$exprs { $param => ..$stats }\"") {
+    val q"new X with T { $self => def m = 42}" = q"new X with T { def m = 42 }"
+    assert(self.show[Structure] === "Term.Param(Nil, Name.Anonymous(), None, None)")
+  }
+
+//  test("4 q\"new { ..$stat } with ..$exprs { $param => ..$stats }\"") {
+//    val q"new {..$stats; val b = 4} with $a {$selff => ..$statz}" = q"new {val a = 2; val b = 4}" // todo fails to compile, uncomment after issue #199 resolved
 //  }
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Why's this test disabled?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 17]
```diff
@@ -306,4 +392,286 @@ class QuasiquoteSuite extends FunSuite {
...
+
+  test("arg\"$expr: _*\"") {
+    val expr = q"foo"
+    assert(arg"$expr: _*".show[Syntax] === "foo: _*")
+  }
+
+  test("arg\"$expr\"") {
+    val expr = q"foo"
+    assert(q"$expr".show[Syntax] === "foo")
+  }
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Why's it say `q"$expr"`?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 18]
```diff
@@ -306,4 +392,286 @@ class QuasiquoteSuite extends FunSuite {
...
+    val tname = t"Y"
+    val tpe2 = t"Z"
+    assert(t"$tpe1 $tname $tpe2".show[Syntax] === "X Y Z")
+  }
+
+  test("t\"(..$atpes) => $tpe\"") {
+    val atpes = List(t"X", t"Y")
+    val tpe = t"Z"
+    assert(t"(..$atpes) => $tpe".show[Syntax] === "(X, Y) => Z")
+  }
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Could we have a test when `..$atpes` are actually `Type.Arg`s here?

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala (hunk 11 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: I don't know exactly, why this changed, but new error message seems to be informative.
For list of types user'd get `found   : scala.collection.immutable.Seq[scala.meta.internal.ast.Type]
 required: scala.meta.internal.ast.Type
       q"expr: ..$tpes"
                 ^
`

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 19]
```diff
@@ -306,4 +392,286 @@ class QuasiquoteSuite extends FunSuite {
...
+    assert(t"$tpe forSome { ..$stats }".show[Syntax] === "X forSome { val a: A; val b: B }")
+  }
+
+  test("t\"$tpe ..@$annots\"") {
+    val tpe = t"X"
+    val annots = List(mod"@a", mod"@b")
+    assert(t"$tpe ..@$annots".show[Syntax] === "X @a @b")
+  }
+
+  test("t\"_ >: $tpeopt <: $tpeopt\"") {
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Why does it say `opt` if you're splicing regular types?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala [hunk 20]
```diff
@@ -306,4 +392,286 @@ class QuasiquoteSuite extends FunSuite {
...
+    assert(tparams(0).show[Syntax] === "X")
+    assert(tparams(1).show[Syntax] === "Y")
+    assert(tpe.show[Syntax] === "(X, Int) => Y")
+  }
+
+  test("2 t\"[..$tparams] => $tpe\"") {
+    val tparams = List(tparam"X", tparam"Y")
+    val tpe = t"Z"
+    assert(t"[..$tparams] => $tpe".show[Syntax] === "[X, Y] => Z")
+  }
```
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: What is the principle according to which some cases get construction/deconstruction tests, and some just get a construction test? :)

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 13 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Changed. And same error in one of the last tests in the file.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 10 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: But isn't `!ahead(token.is[`case`])` going to be false, which will prompt an early exit from the loop?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 14 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Changed.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 9 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: There's the `sepFirst` flag that allows to accommodate this.

File: scalameta/src/main/scala/scala/meta/Trees.scala (hunk 2 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/Trees.scala]: Changed.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 9 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: I'm trying to push unification, because then we'll have less places to change when time will come to make adjustments to our quasiquotes (e.g. when complex ellipses become a thing).

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala (hunk 11 above)
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: Okay, what if `tpes` was not `List(q"T")`, but `List(t"T")`. Would it compile then?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 15 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Done.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 6 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Yeah, sure the suggested change would have to be made to both branches. Then the `unquote` method would go back to its old form - without the parameter and without multiple overloads.

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 16 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Because it does not compile, and waits for issue to be resolved.

Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: I see. Would be cool if you could mention that.

Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Ah I see. Could you move the explaining comment to be above the test?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 17 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Typo, fixed.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 7 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: What if there are multiple statseps here? Also, why statsep and not `;` directly?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 18 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Here they are `Type.Name`, isn't it `Type.Arg`? They both have `t` StringContext for creating.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 8 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Why `CaseIntro` and not `case` like it was before?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 18 above)
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Yeah, sure, I mean could we have `Type.Arg`s that aren't `Type`, so that we really test that function types can accept type args as their arguments?

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 19 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Probably I should add one more comment
`// TODO change to expropt (and test it) after issue #199 resolved`

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 9 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: If unification doesn't work here, then sure no problem - I'm already almost okay with `caseSeparated` going away.

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 20 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: I check all cases for both type of tests in repl. For those who have (or had before) some issues, I explicitly add test.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 3 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Changed.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 6 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: In old form, `unquote` always drives iterator forward. However, sometimes that is not desired behaviour, so I added extra parameter with default value, for the future cases.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 7 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: 1) I thought about that, but it would require some extra changes in code (for cases when scala supports multiple `;`, but this parser doesn't) just for nothing. `;;;;;` is useless action.
2) Because new line might happen inside cases.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 8 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Same reason as with `StatSep` vs `;`. `CaseIntro` is most general form of `case` that can be applied here.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 9 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: I'll try investigate in moving `tokenSeparated` usage back.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 10 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: I still don't understand. `q{case class A}` works.

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala (hunk 11 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: For list of types it says that wants a single type. Agree, so previous message is lying, and need to be changed.

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 12 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Added, typo.

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 16 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Done.

File: tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala (hunk 18 above)
Comment (Dveim) [inline tests/src/test/scala/quasiquotes/QuasiquoteSuite.scala]: Changed test.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 7 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: 1) I agree that this solution looks good enough, but what about consistency with the grammar spec? Scala supports stuff like `foo match { case x => y;;;;;; case z => t;;;; }`, so why not make it support the same in quasiquotes?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 10 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Oh it does. Then it's fine.

File: tests/src/test/scala/quasiquotes/ErrorSuite.scala (hunk 11 above)
Comment (xeno-by) [inline tests/src/test/scala/quasiquotes/ErrorSuite.scala]: I think the new message comes from the fact that you changed `name` to accept ellipses in addition to accepting unquotes. Why was that necessary?

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 6 above)
Comment (xeno-by) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: I suggest we discuss this in voice. Ping me when you're up for that.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 10 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: One of tests didn't pass, and analysis showed that issue can be resolved by adding that checking.

File: scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala (hunk 7 above)
Comment (Dveim) [inline scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala]: Changed `if` to `while`.

**Comment by xeno-by** [comment]
Allright, even though I'm not completely satisfied with `caseClauses`, let's leave that one for future work. LGTM.
